go/fixchain/logger.go

```diff
@@ -182,7 +158,38 @@ func (l *Logger) postChain(p *toPost) {
 			} else {
 				log.Printf(ferr.Error.Error())
 				p.retries--
-				l.postToLog(p)
+				// postToLog() is called in a separate goroutine to avoid
+				// deadlock during retries.  Without the separate goroutine,
+				// deadlock can occur in the following situation:
+				//
+				// Suppose there is only one postServer() goroutine running, and
+				// it is blocked waiting for a toPost on the toPost chan.  A
+				// toPost gets added to the chan, which causes the following to
+				// happen:
+				// - the postServer takes the toPost from the chan.
+				// - the postServer calls l.postChain(toPost), and waits for
+				//   l.postChain() to return before going back to the toPost
+				//   chan for another toPost.
+				// - l.postChain() begins execution.  Suppose the first post
+				//   attempt of the toPost fails for some network-related
+				//   reason.
+				// - l.postChain retries and calls l.postToLog() to queue up the
+				//   toPost to try to post it again.
+				// - l.postToLog() tries to put the toPost on the toPost chan,
+				//   and blocks until a postServer takes it off the chan.
+				// But the one and only postServer is still waiting for
+				// l.postChain (and therefore l.postToLog) to return, and will
+				// not go to take another toPost off the toPost chan until that
+				// happens.
+				// Thus, deadlock.
+				//
+				// Similar situations with multiple postServers can easily be
+				// imagined.
+				l.wg.Add(1)
+				go func() {
+					l.postToLog(p)
+					l.wg.Done()
```

Comment:
The answers to both of your comments lie in the fact that logToPost is not only used in retries, but also when initially queuing chains for logging :)

You need the Add() in postToLog() so that the wg is incremented each time QueueChain() is used to add a chain to the queue of things to post.

This makes a difference to memory consumption because, as it was before, each time a chain was sent to the logger to be queued, QueueChain would call logToPost, which, instead of blocking while waiting for one of the workers to take the chain off the channel, would spin up a new goroutine to block, leaving logToPost and QueueChain to return immediately.  This was resulting in LOADS (10s to 100s of thousands) of new goroutines sitting waiting for a free logger,  each holding a chain.  The latest test run with these changes is showing only ~1.2GB of memory rather than the previous 70GB+ :)